web/src/inbox_ui.js

```diff
@@ -0,0 +1,875 @@
+}
+
+function get_sorted_stream_topic_dict() {
+    const sorted_stream_keys = get_sorted_stream_keys();
+    const sorted_topic_dict = {};
+    for (const sorted_stream_key of sorted_stream_keys) {
+        sorted_topic_dict[sorted_stream_key] = topics_dict[sorted_stream_key];
+    }
+
+    return sorted_topic_dict;
+}
+
+function update_data_and_ui() {
+    pms_dict = {};
+    topics_dict = {};
+    streams_dict = {};
+
+    const unread_pms = unread.get_unread_pm();
+    const unread_pms_count = unread_pms.total_count;
+    const unread_pms_dict = unread_pms.pm_dict;
+
+    const unread_stream_message = unread.get_unread_topics(include_muted);
+    const unread_stream_msg_count = unread_stream_message.stream_unread_messages;
+    const unread_streams_dict = unread_stream_message.stream_count;
+
+    let has_pms_post_filter = false;
+    if (unread_pms_count) {
+        for (const [key, value] of unread_pms_dict) {
+            if (value) {
+                const pm_data = format_pm(key);
+                pms_dict[key] = pm_data;
+                if (!pm_data.is_hidden) {
+                    has_pms_post_filter = true;
+                }
+            }
+        }
+    }
+
+    const has_unread = unread_pms_count + unread_stream_msg_count > 0;
+
+    if (unread_stream_msg_count) {
+        for (const [stream_id, topic_dict] of unread_streams_dict) {
+            const stream_unread = unread.num_unread_for_stream(stream_id);
+            const stream_unread_count = stream_unread.unmuted_count + stream_unread.muted_count;
+            const stream_key = get_stream_key(stream_id);
+            if (stream_unread_count > 0) {
+                topics_dict[stream_key] = {};
+                const stream_data = format_stream(stream_id);
+                for (const [topic, topic_unread_count] of topic_dict) {
+                    if (topic_unread_count) {
+                        const topic_key = get_topic_key(stream_id, topic);
+                        const topic_data = format_topic(stream_id, topic);
+                        topics_dict[stream_key][topic_key] = topic_data;
+                        if (!topic_data.is_hidden) {
+                            stream_data.is_hidden = false;
+                        }
+                    }
+                }
+                streams_dict[stream_key] = stream_data;
+            } else {
+                delete topics_dict[stream_key];
+            }
+        }
+    }
+
+    topics_dict = get_sorted_stream_topic_dict();
+
+    $("#inbox_list").remove();
+
+    $("#inbox_main").append(
+        render_inbox_list({
+            unread_pms_count,
+            has_pms_post_filter,
+            has_unread,
+            pms_dict,
+            topics_dict,
+            streams_dict,
+        }),
+    );
+
+    setTimeout(() => {
+        // We don't want to focus on simplebar ever.
+        $("#inbox_list .simplebar-content-wrapper").attr("tabindex", "-1");
+        revive_current_focus();
+    }, 0);
```

Comment:
Hmm, this is ugly. Would it be possible to avoid needing the `setTimeout` logic here by having the container that gets added/removed during rerendering be something inside the simplebar region?

Reply:
Isn't this inside `update_data_and_ui`, which gets called when changing filters etc.?

Reply:
Oops!

* Add a <div id="inbox-list-inner-wrapper"> inside the simplebar wrapper so that the list can be updated separately.
* Moved this setTimeout to complete_rerender function so that it is no longer a part of update_data_and_ui process.